src/IBM.Cloud.SDK.Core/Util/CredentialUtils.cs

```diff
@@ -0,0 +1,350 @@
+        private static Dictionary<string, List<VcapCredential>> GetVcapServices()
+        {
+            string envServices = Environment.GetEnvironmentVariable(CredentialUtils.vcapServices);
+            if (string.IsNullOrEmpty(envServices))
+            {
+                return null;
+            }
+
+            Dictionary<string, List<VcapCredential>> vcapServices = null;
+
+            try
+            {
+                vcapServices = JsonConvert.DeserializeObject<Dictionary<string, List<VcapCredential>>>(envServices);
+            }
+            catch (Exception e)
+            {
+                Console.WriteLine(string.Format("Error parsing VCAP_SERVICES: {0}", e.Message));
+            }
+
+            return vcapServices;
+        }
+
+        /// <summary>
+        /// A helper method to retrieve the appropriate 'credentials' JSON property value from the VCAP_SERVICES.
+        /// </summary>
+        /// <param name="vcapServices">Dictionary<string, List<VcapCredential>> representing the VCAP_SERVICES</param>
+        /// <param name="serviceName">The name of the service whose credentials are sought</param>
+        /// <param name="plan">The name of the plan for which the credentials are sought, e.g. 'standard', 'beta' etc, may be null</param>
+        /// <returns>The first set of credentials that match the search criteria, service name and plan. May return null</returns>
+        private static VcapCredential GetVcapCredentialsObject(Dictionary<string, List<VcapCredential>> vcapServices, string serviceName, string plan = null)
+        {
+            foreach (KeyValuePair<string, List<VcapCredential>> kvp in vcapServices)
+            {
+                if (kvp.Key.StartsWith(serviceName))
+                {
+                    List<VcapCredential> credentials = kvp.Value;
+                    if (!string.IsNullOrEmpty(plan))
+                    {
+                        foreach (VcapCredential credential in credentials)
+                        {
+                            if (credential.Plan.ToLower() == plan.ToLower())
+                            {
+                                return credential;
+                            }
+                        }
+                    }
+                    else
+                    {
+                        return credentials[0];
+                    }
+                }
+            }
+
+            return null;
+        }
+
+
+        /// <summary>
+        /// Creates a list of files to check for credentials. The file locations are:
+        /// Location provided by user-specified IBM_CREDENTIALS_FILE environment variable
+        /// System home directory (Unix)
+        /// System home directory (Windows)
+        /// Top-level directory of the project this code is being called in
+        /// </summary>
+        /// <returns>List of credential files to check</returns>
+        private static List<List<string>> GetFilesToCheck()
+        {

```

Comment:
It looks like you're potentially reading in each file in the search precedence, and adding its file contents (a list of strings) to a list, then later deciding which list (file contents) to use.
This would actually read in multiple files if more than one of the files actually existed, whereas in the java core we look through the list of files and attempt to read in only the first file that actually exists.

Reply:
You're right - i misread the java core `File` as `FileStream`. I thought the files were already read. Refactored.